### PR TITLE
repart: derive verity salt and uuid from seed

### LINKED ISSUE #28656: repart: set salt & uuid to create reproducible verity partitions
### Component

systemd-repart

### Is your feature request related to a problem? Please describe

Partition definitions which set `Verity=` are not created in a reproducible fashion because the call to _libcryptsetup_ does neither set a [`salt`][cryptsetup-salt] nor a superblock [`uuid`][cryptsetup-uuid]:

https://github.com/systemd/systemd/blob/e3b7fe92e90fd107ea9be30112e00e227ca18501/src/partition/repart.c#L3730-L3740

As a result, even in a hermetic environment, subsequent calls will always result in a different roothash (due to the missing salt) and image (due to the missing uuid in the verity superblock).

[cryptsetup-salt]: https://gitlab.com/cryptsetup/cryptsetup/-/blob/ef46ded7b4e7a95546a2d6d68bfec8abd18f560a/lib/libcryptsetup.h#L535
[cryptsetup-uuid]: https://gitlab.com/cryptsetup/cryptsetup/-/blob/ef46ded7b4e7a95546a2d6d68bfec8abd18f560a/lib/setup.c#L2948

### Describe the solution you'd like

_systemd-repart_ already accepts a [`--seed`][repart-seed] command line argument to generate reproducible partition UUIDs. If `--seed` is omitted, it defaults to the machine ID. The code uses HMAC-SHA256 to derive a UUID from the seed and the partition type:

https://github.com/systemd/systemd/blob/e3b7fe92e90fd107ea9be30112e00e227ca18501/src/partition/repart.c#L4569-L4576

In a similar fashion, `partition_format_verity_hash` could use the seed as an HMAC key to derive a stable salt.

[repart-seed]: https://www.freedesktop.org/software/systemd/man/systemd-repart.html#--seed=

### Describe alternatives you've considered

A possible alternative is adding a new setting to allow configuring the salt explicitly. In that case, it probably makes sense to introduce a new configuration section, e.g., `[Verity]`, to make more parameters configurable (`hash_name`, `salt_size`, ...). As far as I understand, this is something @DaanDeMeyer [also suggests](https://github.com/systemd/systemd/pull/28480#issuecomment-1644998421).

Arguably, this approach allows passing more secure salt values (i.e., more bytes of random data) to dm-verity. Yet, it might still make sense to have a sensible default for a stable uuid and salt value.

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### PR DESCRIPTION
**Pull request by veehaitch**
Fixes #28656

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -3895,17 +3895,56 @@ static int partition_format_verity_hash(
...
+                                        .flags = CRYPT_VERITY_CREATE_HASH,
+                                        .hash_name = "sha256",
+                                        .hash_type = 1,
+                                        .data_block_size = context->sector_size,
+                                        .hash_block_size = context->sector_size,
+                                        .salt_size = 32,
+                                });
+        } else {
+                sd_id128_t uuid;
+                char uuid_string[37];
```
Comment (bluca) [inline src/partition/repart.c]: SD_ID128_UUID_STRING_MAX

File: src/partition/repart.c [hunk 2]
```diff
@@ -3895,17 +3895,56 @@ static int partition_format_verity_hash(
...
+                const char salt_token[] = "verity-salt";
+
+                hmac_sha256(context->seed.bytes, sizeof(context->seed.bytes), salt_token, strlen(salt_token), salt.md);
+
+                r = derive_uuid(context->seed, "verity-uuid", &uuid);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to acquire verity uuid: %m");
+
+                /* crypt_format expects a UUID string */
+                r = snprintf(uuid_string, sizeof(uuid_string), SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(uuid));
```
Comment (bluca) [inline src/partition/repart.c]: sd_id128_to_uuid_string()

Comment (veehaitch) [reply src/partition/repart.c]: Thanks! Guess I can use `SD_ID128_TO_UUID_STRING(uuid)` directly

**Review by DaanDeMeyer — COMMENTED** [review]
On the phone but there's a function where we derive all the uuids. We should probably derive the uuid anf salt for verity in that function as well.

**Review by bluca — APPROVED** [review]

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Alright, got to check, we have `context_acquire_partition_uuids_and_labels()` where we should derive the UUIDs and the salt. You can store the uuid and salt in two new fields in the `Partition` struct.

Also, there's no need to conditionalize the behavior on `arg_randomize`. When `arg_randomize` is set, we already use a random seed and so the derived UUID and salt will be random as well, so we can just use one codepath instead of two.

**Comment by veehaitch** [comment]
> Also, there's no need to conditionalize the behavior on `arg_randomize`. When `arg_randomize` is set, we already use a random seed and so the derived UUID and salt will be random as well, so we can just use one codepath instead of two.

Yes, that's what I wanted to do in the first place. But then, I also wanted to make a difference between `--seed=random` on one hand and an explicit and an implicit seed on the other hand to not reduce the entropy of the salt unnecessarily if "random" was requested.

**Comment by DaanDeMeyer** [comment]
> Yes, that's what I wanted to do in the first place. But then, I also wanted to make a difference between --seed=random on one hand and an explicit and an implicit seed on the other hand to not reduce the entropy of the salt unnecessarily if "random" was requested.

I would stick with the one logic approach I suggested. That way the verity uuid and salt are always deterministically derived of the seed, even if the seed itself is random. This matches nicely with how we calculate uuids and salts for all other partitions.

**Comment by veehaitch** [comment]
> I would stick with the one logic approach I suggested. 

Got it.

> Alright, got to check, we have `context_acquire_partition_uuids_and_labels()` where we should derive the UUIDs and the salt. You can store the uuid and salt in two new fields in the `Partition` struct.

I am not too familiar with the systemd-repart codebase, still, are you sure that's the right place? I mean, we only need the salt and the superblock uuid if we have a `VERITY_HASH` partition.

**Comment by DaanDeMeyer** [comment]
> I am not too familiar with the systemd-repart codebase, still, are you sure that's the right place? I mean, we only need the salt and the superblock uuid if we have a VERITY_HASH partition.

Yes, check the existing logic for the luks UUID in context_acquire_partition_uuids_and_labels(), it is only allocated if encryption is actually enabled. We'd do the same for the verity UUID and salt. Check if the partition is a verity hash partition and if so, populate new verity_uuid and verity_salt fields in the Partition struct. And then use those in format_verity_hash(). Not sure why @poettering chose to do it this way, but let's follow the same approach.

File: src/partition/repart.c [hunk 3]
```diff
@@ -208,6 +208,8 @@ typedef struct Partition {
         bool new_uuid_is_set;
         char *current_label, *new_label;
         sd_id128_t fs_uuid, luks_uuid;
+        sd_id128_t verity_uuid;
```
Comment (daandemeyer) [inline src/partition/repart.c]: Please merge this with the previous declaration

File: src/partition/repart.c [hunk 4]
```diff
@@ -4866,6 +4869,20 @@ static int context_acquire_partition_uuids_and_labels(Context *context) {
...
                 }
 
+                /* Derive the verity salt and verity superblock UUID from the seed to keep them reproducible */
+                if (p->verity == VERITY_HASH) {
+                        const char salt_token[] = "verity-salt";
+                        hmac_sha256(context->seed.bytes,
+                                    sizeof(context->seed.bytes),
+                                    salt_token,
+                                    strlen(salt_token),
+                                    p->verity_salt);
```
Comment (daandemeyer) [inline src/partition/repart.c]: Can you add `derive_salt()` similar to `derive_uuid()` for this?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks great! Two small comments

File: src/partition/repart.c [hunk 5]
```diff
@@ -4866,6 +4869,20 @@ static int context_acquire_partition_uuids_and_labels(Context *context) {
                                 return r;
                 }
 
+                /* Derive the verity salt and verity superblock UUID from the seed to keep them reproducible */
+                if (p->verity == VERITY_HASH) {
+                        const char salt_token[] = "verity-salt";
```
Comment (keszybz) [inline src/partition/repart.c]: We don't need to define a variable for this. Please just use a define:
```C
#define VERITY_SALT "verity-salt"
```
and then use `VERITY_SALT` and `strlen(VERITY_SALT)` in the call.

Comment (daandemeyer) [reply src/partition/repart.c]: Even the define is not needed if we add `derive_salt()` like I proposed.

**Comment by poettering** [comment]
looks conceptually good to me

**Comment by poettering** [comment]
lgtm

**Review by bluca — APPROVED** [review]

**Review by keszybz — APPROVED** [review]

**Pull request merged**